### ISSUE TITLE
Uneven shard distribution and download traffic

### ISSUE THREAD

**Issue by littleskunk** [Storj/bridge#344]
### Package Versions

Replace the values below using the output from `npm list storj-bridge`.

```
root@storj:~# storj --version
Storjcli: 4.0.0 | Core: 6.0.15
```

Replace the values below using the output from `node --version`.

```
v6.9.4
```

### Expected Behavior

Please describe the program's expected behavior. Include an example of your
usage code in the back ticks below if applicable.

I uploaded a tiny file. Farmer `e6ac6e2f73990df43daf3ca61f97e9210b7bbe80` received the shard and 5 other farmer received mirrors. If I try to download the file I would expect a even distribution beween all 6 farmer. They should all get a few download bytes.
```
root@storj:~# storj upload-file 57449d89dc436e0a48f041aa .storjcli/upload
 [Tue Jan 31 2017 22:22:10 GMT+0100 (CET)] [info]   1 file(s) to upload.
 [...]  > Enter your passphrase to unlock your keyring  >  ********

 [Tue Jan 31 2017 22:22:14 GMT+0100 (CET)] [info]   Generating encryption key...
 [Tue Jan 31 2017 22:22:14 GMT+0100 (CET)] [warn]   [ upload ] Already exists in bucket. Uploading to (2017-01-31T21;22;14.937Z)-upload
 [Tue Jan 31 2017 22:22:14 GMT+0100 (CET)] [info]   [ (2017-01-31T21;22;14.937Z)-upload ] Creating storage token... (retry: 0)
 [Tue Jan 31 2017 22:22:15 GMT+0100 (CET)] [info]   Encrypting file ".storjcli/upload"
 [Tue Jan 31 2017 22:22:15 GMT+0100 (CET)] [info]   [ (2017-01-31T21;22;14.937Z)-upload ] Encryption complete
 [Tue Jan 31 2017 22:22:15 GMT+0100 (CET)] [info]   [ (2017-01-31T21;22;14.937Z)-upload ] Storing file, hang tight!
 [Tue Jan 31 2017 22:22:15 GMT+0100 (CET)] [info]   Creating file staging frame
 [Tue Jan 31 2017 22:22:16 GMT+0100 (CET)] [info]   Trying to upload shard /opt/tmp/9d7bfdf84b40 index 0
 [Tue Jan 31 2017 22:22:16 GMT+0100 (CET)] [info]   Hash for this shard is: e78cbd02eacf2d603887687a2bfe3657c9353372
 [Tue Jan 31 2017 22:22:16 GMT+0100 (CET)] [info]   Audit generation for shard done.
 [Tue Jan 31 2017 22:22:16 GMT+0100 (CET)] [info]   Waiting on a storage offer from the network...
 [Tue Jan 31 2017 22:22:16 GMT+0100 (CET)] [info]   Querying bridge for contract for e78cbd02eacf2d603887687a2bfe3657c9353372 (retry: 0)
 [Tue Jan 31 2017 22:22:17 GMT+0100 (CET)] [info]   Contract negotiated with: {"userAgent":"6.0.15","protocol":"1.1.0","address":"003s.storjlt.xyz","port":4601,"nodeID":"e6ac6e2f73990df43daf3ca61f97e9210b7bbe80","lastSeen":1485897735360}
 [Tue Jan 31 2017 22:22:18 GMT+0100 (CET)] [info]   Shard transfer completed! 0 remaining...
 [Tue Jan 31 2017 22:22:18 GMT+0100 (CET)] [info]   Transfer finished, creating entry.. (retry: 0)
 [Tue Jan 31 2017 22:22:18 GMT+0100 (CET)] [info]   sending exchange report
 [Tue Jan 31 2017 22:22:18 GMT+0100 (CET)] [info]   [ (2017-01-31T21;22;14.937Z)-upload ] Cleaning up...
 [Tue Jan 31 2017 22:22:18 GMT+0100 (CET)] [info]   [ (2017-01-31T21;22;14.937Z)-upload ] Finished cleaning!
 [Tue Jan 31 2017 22:22:18 GMT+0100 (CET)] [info]   [ (2017-01-31T21;22;14.937Z)-upload ] Encryption key saved to keyring.
 [Tue Jan 31 2017 22:22:18 GMT+0100 (CET)] [info]   [ (2017-01-31T21;22;14.937Z)-upload ] File successfully stored in bucket.
 [Tue Jan 31 2017 22:22:18 GMT+0100 (CET)] [info]   Name: (2017-01-31T21;22;14.937Z)-upload, Type: application/octet-stream, Size: 1280 bytes, ID: 76cf610ece582d974724db91
 [Tue Jan 31 2017 22:22:18 GMT+0100 (CET)] [info]   1 of 1 files uploaded
 [Tue Jan 31 2017 22:22:18 GMT+0100 (CET)] [info]   Done.
root@storj:~# storj list-mirrors 57449d89dc436e0a48f041aa 76cf610ece582d974724db91
 [Tue Jan 31 2017 22:22:47 GMT+0100 (CET)] [info]
 [Tue Jan 31 2017 22:22:47 GMT+0100 (CET)] [info]   Established
 [Tue Jan 31 2017 22:22:47 GMT+0100 (CET)] [info]   -----------
 [Tue Jan 31 2017 22:22:47 GMT+0100 (CET)] [info]   Shard: 0
 [Tue Jan 31 2017 22:22:47 GMT+0100 (CET)] [info]   Hash: e78cbd02eacf2d603887687a2bfe3657c9353372
 [Tue Jan 31 2017 22:22:47 GMT+0100 (CET)] [info]       storj://47.184.44.104:53261/e7c0f0d9ace3573ce1ef284ad646e75326e3854b
 [Tue Jan 31 2017 22:22:47 GMT+0100 (CET)] [info]       storj://80.76.183.72:9101/e6b872ad219ecf753cc9228a550bdd234f6745e0
 [Tue Jan 31 2017 22:22:47 GMT+0100 (CET)] [info]       storj://client015.storj.dk:15056/e7871598b710ccf94f6630b5b9d9caed54f4d607
 [Tue Jan 31 2017 22:22:47 GMT+0100 (CET)] [info]       storj://152.231.100.127:24054/e79aaac645c571dd8178b68065cfd5c40693f328
 [Tue Jan 31 2017 22:22:47 GMT+0100 (CET)] [info]       storj://92.52.142.15:27832/e7e466fc25166bb2e7b85b740d4041083f1c1b3f
 [Tue Jan 31 2017 22:22:47 GMT+0100 (CET)] [info]
 [Tue Jan 31 2017 22:22:47 GMT+0100 (CET)] [info]   Available
 [Tue Jan 31 2017 22:22:47 GMT+0100 (CET)] [info]   ---------
 [Tue Jan 31 2017 22:22:47 GMT+0100 (CET)] [info]   Shard: 0
 [Tue Jan 31 2017 22:22:47 GMT+0100 (CET)] [info]   Hash: e78cbd02eacf2d603887687a2bfe3657c9353372
 [Tue Jan 31 2017 22:22:47 GMT+0100 (CET)] [info]       storj://89.176.149.45:37700/e7028eddc21f9ceaa4de94e80290f23382a49dcf
 [Tue Jan 31 2017 22:22:47 GMT+0100 (CET)] [info]       storj://client014.storj.dk:15050/e7c0f7c62b37d2fcf364dce4ee181e5c917d75ca
 [Tue Jan 31 2017 22:22:47 GMT+0100 (CET)] [info]       storj://client012.storj.dk:15056/e6ef38ec5173863cbf57ed87138449a81f25c345
 [Tue Jan 31 2017 22:22:47 GMT+0100 (CET)] [info]       storj://client021.storj.dk:15049/e672512fd7ac2091abe95e90bec433696844a083
 [Tue Jan 31 2017 22:22:47 GMT+0100 (CET)] [info]       storj://client015.storj.dk:15051/e697688f4d89b82f55105c9ee0bc84592cf16e13
```

### Actual Behavior

Please describe the program's actual behavior. Please include any stack traces
or log output in the back ticks below.

5 times in a row I downloaded from the same farmer `e7c0f0d9ace3573ce1ef284ad646e75326e3854b`. I did the same test 3 times with new uploaded files. The download traffic goes always to the same mirror farmer. The original farmer and the other 4 mirror farmer will get 0 traffic.
```
 root@storj:~# storj get-pointers 57449d89dc436e0a48f041aa 76cf610ece582d974724db91
 [Tue Jan 31 2017 22:31:08 GMT+0100 (CET)] [info]   Listing pointers for shards 0 - 0
 [Tue Jan 31 2017 22:31:08 GMT+0100 (CET)] [info]   -----------------------------------------
 [Tue Jan 31 2017 22:31:08 GMT+0100 (CET)] [info]
 [Tue Jan 31 2017 22:31:08 GMT+0100 (CET)] [info]   Index:  0
 [Tue Jan 31 2017 22:31:08 GMT+0100 (CET)] [info]   Hash:   e78cbd02eacf2d603887687a2bfe3657c9353372
 [Tue Jan 31 2017 22:31:08 GMT+0100 (CET)] [info]   Token:  876841e23eb5900ae5244b06b5f09706ab36584e
 [Tue Jan 31 2017 22:31:08 GMT+0100 (CET)] [info]   Farmer: storj://47.184.44.104:53261/e7c0f0d9ace3573ce1ef284ad646e75326e3854b
 [Tue Jan 31 2017 22:31:08 GMT+0100 (CET)] [info]
```

### Steps to Reproduce

Please include the steps the reproduce the issue, numbered below. Include as
much detail as possible.

1. Upload a new file.
2. Wait until the bridge created 5 mirrors.
3. Download the files multiple times and track the nodeIDs.

**Comment by braydonf** [comment]
It's going to select the pointer that is returned first. I don't see an issue here, as the one that responds most quickly is also likely best able to handle the download.

**Comment by littleskunk** [comment]
The result will be a centralized USA network. Setup a datacenter on the same network as the bridge server and you will get all shards and the download traffic.

The available download bandwith will be ignored. A random distribution could need 5 more seconds but at the end 5 times more bandwith.

**Comment by littleskunk** [comment]
Same problem with shard distribution: https://github.com/Storj/bridge/blob/6e4651ccf8e34b6d7c4afe75b896b5ff5cbbc393/lib/server/routes/reports.js#L134

> But sometimes things don’t go as planned, like in the RSA backdoor instance. In the case of such a nightmare scenario, a distributed network has a key advantage over traditional storage systems: the attacker has to know where the shards are stored in order to retrieve and decrypt them. Those shard locations can be secured separately from the keys themselves, and are essential to retrieving a file. This is where Storj’s security proposition becomes highly relevant again.
> http://blog.storj.io/post/145305561698/how-storj-increases-object-storage-security

At the moment implemented is a very unsecure shard distribution. As an attacker I can call api.storj.io/contacts and order it by response time. This way I can reduce the number of farmer. I don't need to ask 4000 farmer. A contract PUBLISH will allow more than 700 farmer to send an OFFER. I can group the farmer by nodeID. Lets use the first byte of the nodeID to build 16 groups. If I ask the 6 low response time farmer of each group I can get the complete file in a short time.

Second problem:
Lets setup a datacenter located at the bridge network. Run 6 farmer with close nodeIDs pointing all to the same kfs db. I only have to store the shard once but I will get the payment for all 5 mirrors as well. Audit is possible for all 6 shards.
What is going to happen if my HDD crashes? The bridge will lose the complete file and there is no chance to recover it.

Response time is not good for shard distribution.

**Comment by littleskunk** [comment]
One proof of my theory you can see in #391. The same node send 4 OFFER and gets 4 shards. This farmer will store the shard only once. High risk of loosing the file. Even if we add a rule to filter unique nodeIDs that doesn't solve the distribution problem. That one mirror farmer has a timeoutRate of 2% while other farmer with 0% are on the queue.
The problem is that the bridge will start the mirror process as soon as the first OFFER is there. -> The bridge is selecting by response time and vulnerable that way.

Even if that one farmer had 0% timeoutRate at the moment the shard was uploaded. I would expect a random selection over all 0% timeoutRate farmer.

**Comment by littleskunk** [comment]
> I'm still getting a ton of slow nodes from all over the world (some of them I'm even uploading @ 10<->50kbit/s to.. ;))
Of the 8½ hours to upload the 1TB file, 98% of the shards came trough in the first 3 hours or so.. the remaining ~5 hours was due to slow nodes.
If the bridge are choosing farmers with the lowest ping, you could still end up with slow nodes, no doubt about that.
IMHO a better way would be to use the exchange reports to find nodes that has way better upload / download speed between renters IP and IP's of those farmers that returns offers to the bridge.
The reason I'm suggestion this are because many ISP/DC are prioritizing their bandwidth.
eg. if I'm trying to send a shard to a farmer in .jp the odds are pretty high for slow speed to this particular farmer, since I'm located in Denmark (which are pretty far away from .jp = high latency)
also, I've seen slow data transfers between to ISP's which where located in the same country. All the traffic between them had to go via 100GBit/s but in another country, which resulted in very high ping.
What I'm trying to say is even if you have a fast pipe like 1Gb/10Gb/100Gb/s you have absolutely no control over how the ISP/DC prioritize their traffic, so uploading/downloading to your nextdoor neighbor could go as slow as he** because he/she are using a different ISP that might be exchanging traffic using a 3. party ISP/DC which again are using BGP4 to route upload traffic though one pipe and download though another pipe. 
Therefor I would suggest using the exchange reports to make some kind of a "speed-map" to figure out which farmers can provide renters with a good upload/download speed using both IP addresses. The "speed-map" could also be used when farmers has to establish mirrors to other farmers.
The "speed-map" would be used by the bridge (using mongodb) so a fast decision between the incoming offers can be made.
The so called "speed-map" would have to be build over time, but to help build the map storjshare-daemon or the GUI could have a "help build speed map" feature of some sort, which would randomly select 100 nodes and run a light speed test in both directions, that both testers would report to the bridge.
To extend the "speed-map" idea a bit, storj could also be using BGP4 ASN's as a form of "zoning" ranges of IP's which every single ISP/DC in the world are using. This way we could be 100% sure not a single shard would hit the same farmer twice.
>
> Th3Van on rocket chat channel stress-testers

**Comment by littleskunk** [comment]
Low ping USA farmer having problems with OutOfMemory exceptions while my farmer has no traffic and is running for days. They are reporting about 1.5GB per farmer. My farmer is running with only 200MB.

It could be a memory leak but nothing I can reproduce on my system. Maybe someone can help me out with remote access to a GUI node. That would allow me to search for a possible memor leak.

On the renter side I would expect slow download speed as reported by Th3Van. The low ping farmer is getting more and more open downloads. There is no limit. At some point the farmer will crash with a OutOfMemory exception or to many open files.

**Comment by littleskunk** [comment]
The attack is real now. First farmer are running more than 1000 nodes. They see the localhost mirror creation and are now adding even more farmer to get more used space without having the traffic.

Now I would expect shard loss. Reed solomon will decrease the risk but is not able to recover more than 1000 nodes going offline at the same second.

**Comment by tempestb** [comment]
Suggestion -

Right now, the Bridge will send the first shard to the node that responds the fastest with an OFFER.   Assuming this is done for performance reasons, I can agree with this approach, however...

With mirrors, the OFFERS are ordered by response time, and the top 5 are give a shard.   This would be better if it instead chose the mirror nodes based on hash distance.  That way, all nodes have an equal chance of receiving a mirror shard.  This would more evenly distribute shards and not just favor those who are geographically closest to the Bridge (Or Bridges in the future)

**Comment by braydonf** [comment]
There will be load balancing on the initial "broadcast" to farmers in https://github.com/Storj/integration/pull/5, e.g. farmers with a score above the threshold are all in the *same* pool e.g. not much difference between 2000 milliseconds and 20 milliseconds. This should give a much better distribution.

**Comment by littleskunk** [comment]
Outdated. As farmer I would love to see a perfect shard distribution but the main point for this issue was the renter point of view. Shard distribution is secure and the upload is fast.

Shard downloads are a littlebit complicated. At the moment bridge will ask the farmer with the latest LastSeen. The request itself will update LastSeen. If the renter requests the same shard multiple times in a short timeframe all requests will be redirected to the same farmer. At the end only low speed transfer. This could be an issue for public file sharing.
On the other side it should work for random download requests. A random farmer will have the latest LastSeen and on the next download request another farmer will win. High Transfer speed.

Anyway at the moment I can't proof the public file sharing theory. Time to close the issue and open a new one if needed.

**Comment by braydonf** [comment]
I was thinking about this the other day, and I think we can apply the same logic that we did for uploads for downloads. Rather than always returning the first reply, wait some number of milliseconds and respond with a random selection to give better distribution. As many of these requests are asynchronous, the slight delay should not be noticed client side. We can track this with a separate issue, as well as other improvements such as caching already existing download tokens from previous requests (https://github.com/Storj/bridge/issues/493). Now that tokens are persistent on restarts for farmers, that should be possible.
